Google.Api.Gax.Grpc/GrpcChannelOptions.cs

```diff
@@ -1,45 +1,282 @@
+        {
+            EnableServiceConfigResolution = enableServiceConfigResolution;
+            KeepAliveTime = keepAliveTime;
+            PrimaryUserAgent = primaryUserAgent;
+            MaxSendMessageSize = maxSendMessageSize;
+            MaxReceiveMessageSize = maxReceiveMessageSize;
+            CustomOptions = customOptions ?? s_emptyOptions;
+        }
+
+        /// <summary>
+        /// If non-null, explicitly enables or disables service configuration resolution.
+        /// </summary>
+        public bool? EnableServiceConfigResolution { get; }
+
+        /// <summary>
+        /// If non-null, explicitly specifies the keep-alive period for the channel.
+        /// </summary>
+        public TimeSpan? KeepAliveTime { get; }
+
+        /// <summary>
+        /// If non-null, explicitly specifies the primary user agent for the channel.
+        /// </summary>
+        public string PrimaryUserAgent { get; }
+
+        /// <summary>
+        /// If non-null, explicitly specifies the maximum size in bytes that can be sent from the client, per request.
+        /// </summary>
+        public int? MaxSendMessageSize { get; }
+
+        /// <summary>
+        /// If non-null, explicitly specifies the maximum size in bytes that can be received from the client, per response.
+        /// </summary>
+        public int? MaxReceiveMessageSize { get; }
+
+        /// <summary>
+        /// Immutable list of custom options. This is never null, but may be empty.
+        /// </summary>
+        public IReadOnlyList<CustomOption> CustomOptions { get; }
+
+        // TODO: Add static factory methods? The chaining (starting from Empty) is probably simpler.
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="userAgent"></param>
+        /// <returns></returns>
+        public GrpcChannelOptions WithPrimaryUserAgent(string userAgent) =>
+            MergedWith(new GrpcChannelOptions(primaryUserAgent: GaxPreconditions.CheckNotNull(userAgent, nameof(userAgent))));
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="enableServiceConfigResolution"></param>
+        /// <returns></returns>
+        public GrpcChannelOptions WithEnableServiceConfigResolution(bool enableServiceConfigResolution) =>
+            MergedWith(new GrpcChannelOptions(enableServiceConfigResolution: enableServiceConfigResolution));
+
         /// <summary>
-        /// "After a duration of this time the client/server pings its peer to see if the
-        /// transport is still alive. Int valued, milliseconds."
-        /// Required for any channel using a streaming RPC, to ensure an idle stream doesn't
-        /// allow the TCP connection to be silently dropped by any intermediary network devices.
-        /// 60 second keepalive time is reasonable. This will only add minimal network traffic,
-        /// and only if the channel is idle for more than 60 seconds.
+        /// 
```

Comment:
Is `summary` deliberately empty for all these `With` methods?

Reply:
For this commit, yes. I didn't want to write a load of docs and then have to change them.